packages/core/contracts/financial-templates/optimistic-distributor/OptimisticDistributor.sol

```diff
@@ -0,0 +1,425 @@
+     * @dev Calling this for unresolved proposals will revert. Both accepted and rejected distribution
+     * proposals will be deleted from storage.
+     */
+    function executeDistribution(uint256 proposalIndex) external nonReentrant() {
+        // All valid proposals should have non-zero proposal timestamp.
+        Proposal memory proposal = proposals[proposalIndex];
+        require(proposal.timestamp != 0, "Invalid proposalIndex");
+
+        // Append proposal index to ancillary data.
+        Reward memory reward = rewards[proposal.rewardIndex];
+        bytes memory ancillaryData = _appendProposalIndex(proposalIndex, reward.customAncillaryData);
+
+        // Get resolved price. Reverts if the request is not settled or settleable.
+        int256 resolvedPrice =
+            optimisticOracle.settleAndGetPrice(reward.priceIdentifier, proposal.timestamp, ancillaryData);
+
+        // Transfer rewards to MerkleDistributor for accepted proposal.
+        if (resolvedPrice == 1e18) {
+            reward.rewardToken.safeApprove(address(merkleDistributor), reward.maximumRewardAmount);
+            merkleDistributor.setWindow(
+                reward.maximumRewardAmount,
+                address(reward.rewardToken),
+                proposal.merkleRoot,
+                proposal.ipfsHash
+            );
+            emit RewardDistributed(
+                reward.sponsor,
+                reward.rewardToken,
+                proposal.rewardIndex,
+                proposalIndex,
+                reward.maximumRewardAmount,
+                proposal.merkleRoot,
+                proposal.ipfsHash
+            );
+        } else emit ProposalRejected(proposal.rewardIndex, proposalIndex);
+
+        // Delete resolved proposal from storage. This also avoids double-spend for approved proposals.
+        delete proposals[proposalIndex];
+        emit ProposalDeleted(proposal.rewardIndex, proposalIndex);
+    }
+
+    /**
+     * @notice Allows any caller to delete distribution that was rejected by the Optimistic Oracle.
+     */
+    function deleteRejectedDistribution(uint256 proposalIndex) external nonReentrant() {}
+
+    /********************************************
+     *          MAINTENANCE FUNCTIONS           *
+     ********************************************/
+
+    /**
+     * @notice Sets address of MerkleDistributor contract that will be used for rewards distribution.
+     * MerkleDistributor address can only be set once.
+     * @dev It is expected that the deployer first deploys MekleDistributor contract and transfers its ownership to
+     * the OptimisticDistributor contract and then calls `setMerkleDistributor` on the OptimisticDistributor pointing
+     * on now owned MekleDistributor contract.
+     * @param _merkleDistributor Address of the owned MerkleDistributor contract.
+     */
+    function setMerkleDistributor(MerkleDistributor _merkleDistributor) external nonReentrant() {
```

Comment:
I'm curious why you can't pass in the `merkleDistributor` address in the constructor?

Reply:
`merkleDistributor` should be owned by the OptimisticDistributor that is not yet deployed.

Should I create separate creator contract similarly as done in LSP taking ownership on long/short tokens?

Reply:
we can create a factory but leave that for a follow on PR. what you have now looks great.